packages/api/endpoints/rules.js

```diff
@@ -55,24 +59,32 @@ async function get(req, res) {
  * @returns {Promise<Object>} the promise of express response object
  */
 async function post(req, res) {
-  const data = req.body;
-  const name = data.name;
-
-  const model = new models.Rule();
-
   try {
-    await model.get({ name });
-    return res.boom.conflict(`A record already exists for ${name}`);
-  } catch (e) {
-    if (e instanceof RecordDoesNotExist) {
-      const record = await model.create(data);
-
-      if (inTestMode()) {
-        await addToLocalES(record, indexRule);
+    const data = req.body;
+    const name = data.name;
+
+    const model = new models.Rule();
+
+    try {
+      await model.get({ name });
+      return res.boom.conflict(`A record already exists for ${name}`);
+    } catch (e) {
+      if (e instanceof RecordDoesNotExist) {
+        const record = await model.create(data);
+
+        if (inTestMode()) {
+          await addToLocalES(record, indexRule);
+        }
+        return res.send({ message: 'Record saved', record });
       }
-      return res.send({ message: 'Record saved', record });
+      throw e;
     }
-    throw e;
+  } catch (e) {
+    if (e.name === 'SchemaValidationError') {
+      return res.boom.badRequest(e.message);
+    }
+    log.error('Error occurred while trying to create rule:', e);
+    return res.boom.badImplementation(e.message);
```

Comment:
Returning `badImplementation` obscures the errors from the dashboard since for some reason it ignores the message. We should change this to another type or just throw the error.

Reply:
Right - that's why I'm adding the `log.error` statements so that people will at least be able to see what went wrong in the Cloudwatch logs. I think `badImplementation` is what we want though, since a generic 500 error is appropriate when something unspecific went wrong on the server

Reply:
The problem is the dashboard is not seeing the error and is unable to display it on the front end. We don't want to force people to have to go to the logs and see it when they should be able to see error details on the dashboard.

So I don't know why this one is the only one that doesn't send through the message: https://hapi.dev/module/boom/api/?v=9.1.0#boombadimplementationmessage-data---alias-internal

But maybe we could use expectationFailed to pass the message back to the dashboard? https://hapi.dev/module/boom/api/?v=9.1.0#boomexpectationfailedmessage-data

Reply:
Expectation failed is a specific error code for when the conditions in an Expect header cannot be met

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/417

The closest thing to a "generic" error which would return the message to the user is 400/badRequest, but that is specifically for when the user's request is something invalid that the server can't handle. But we could get to this line because collection.create has a bug and throws an error, which would not be a bad request but a general server error. I think that's why Boom obscures the error message from the user for 5xx errors - because we don't want to leak implementation details in our error responses if the fault is in our code.

The genesis of this ticket was that Brian B didn't get any meaningful error from the API when creating a collection and he had an invalid `item.granuleIdExtraction`:

https://github.com/nasa/cumulus/blob/master/packages/api/models/collections.js#L43 

That throws a general Error, which is handled by the endpoint as returning a 500 error, but really it was a bad request. So maybe a better solution would be ensuring that anything during the record creation process which could be caused by a bad request is properly caught and returned to the user as a 400/badRequest? And then we keep 500/badImplementation with the logging for unexpected errors.

Reply:
I'll have to test this to see what errors we're not handling then because I have seen this on the dashboard. i think then we get into a situation where we have to specifically handle every case, which I guess is not a bad thing. 

But it's making me feel like we need some kind of model error handler utility function where we could DRY 
```
if (e.name === 'SchemaValidationError') {
      return res.boom.badRequest(e.message);
    }
    log.error('Error occurred while trying to create rule:', e);
    return res.boom.badImplementation(e.message);
```
and not have this same code throughout our model endpoints.

Reply:
What do you think about this?

019ac966f29242c9cf078bcf7ade78c125d9e1c2

It ensures that all the error cases are handled by the endpoint. It also didn't make sense to me that the model should throw a  "BadRequestError" error, so I created more specific error classes.

This handling would be easily to replicate for the other POST endpoints.

Reply:
Yep I like that better. Thanks!

Reply:
Instead of using `badImplementation` which obscures the error from the dashboard, how about `serverUnavailable`?

Reply:
It wouldn't make a difference:

> All 500 errors hide your message from the end user.

https://hapi.dev/module/boom/api/?v=9.1.0#http-5xx-errors

Reply:
Yep, makes sense. I'd keep it as `badImplementation` then, and the dashboard code has to address things properly. Not sure if it's the 500 responses that cause the dashboard to fail silently, but that's a matter of fixing the dashboard, I believe, not with changing the 500 response itself.

Reply:
It's the 500 errors that are hiding the error message from the user. For 500 errors, the dashboard will only get a response that looks like:

```json
{
  "status": 500,
  "message": "Internal server error"
}
```

That's why we're trying to catch and handle any possible errors that shouldn't return a 500 response so the dashboard will get a helpful error message.

Reply:
Yep, I think we're in agreement here. I suppose I'm trying to clarify that sending a 500 response in the absence of being able to provide something more meaningful is perfectly valid.

Reply:
OK cool. Sorry if I misunderstood 👍